src/mpl/src/gpu/mpl_gpu_cuda.c

```diff
@@ -0,0 +1,357 @@
+    CUDA_ERR_CHECK(ret);
+
+  fn_exit:
+    return MPL_GPU_SUCCESS;
+  fn_fail:
+    return MPL_GPU_FAILURE;
+}
+
+int MPL_gpu_memcpy_h2d_async(void *dst, const void *src, size_t count, MPL_gpu_stream_t stream)
+{
+    cudaError_t ret;
+    ret = cudaMemcpyAsync(dst, src, count, cudaMemcpyHostToDevice, stream);
+    CUDA_ERR_CHECK(ret);
+
+  fn_exit:
+    return MPL_GPU_SUCCESS;
+  fn_fail:
+    return MPL_GPU_FAILURE;
+}
+
+int MPL_gpu_memcpy_d2h_async(void *dst, const void *src, size_t count, MPL_gpu_stream_t stream)
+{
+    cudaError_t ret;
+    ret = cudaMemcpyAsync(dst, src, count, cudaMemcpyDeviceToHost, stream);
+    CUDA_ERR_CHECK(ret);
+
+  fn_exit:
+    return MPL_GPU_SUCCESS;
+  fn_fail:
+    return MPL_GPU_FAILURE;
+}
+
+int MPL_gpu_memcpy_d2d_async(void *dst, const void *src, size_t count, MPL_gpu_stream_t stream)
+{
+    cudaError_t ret;
+    ret = cudaMemcpyAsync(dst, src, count, cudaMemcpyDeviceToDevice, stream);
+    CUDA_ERR_CHECK(ret);
+
+  fn_exit:
+    return MPL_GPU_SUCCESS;
+  fn_fail:
+    return MPL_GPU_FAILURE;
+}
+
+int MPL_gpu_ipc_get_mem_handle(MPL_gpu_ipc_mem_handle_t * h_mem, void *ptr)
+{
+    cudaError_t ret;
+    ret = cudaIpcGetMemHandle(h_mem, ptr);
+    CUDA_ERR_CHECK(ret);
+
+  fn_exit:
+    return MPL_GPU_SUCCESS;
+  fn_fail:
+    return MPL_GPU_FAILURE;
+}
+
+int MPL_gpu_ipc_open_mem_handle(void *ptr, MPL_gpu_ipc_mem_handle_t * h_mem)
+{
+    cudaError_t ret;
+    ret = cudaIpcOpenMemHandle(ptr, h_mem, cudaIpcMemLazyEnablePeerAccess);
+    CUDA_ERR_CHECK(ret);
+
+  fn_exit:
+    return MPL_GPU_SUCCESS;
+  fn_fail:
+    return MPL_GPU_FAILURE;
+}
+
+int MPL_gpu_ipc_close_mem_handle(void *ptr)
+{
+    cudaError_t ret;
+    ret = cudaIpcCloseMemHandle(ptr);
+    CUDA_ERR_CHECK(ret);
+
+  fn_exit:
+    return MPL_GPU_SUCCESS;
+  fn_fail:
+    return MPL_GPU_FAILURE;
+}
+
+int MPL_gpu_ipc_get_event_handle(MPL_gpu_ipc_event_handle_t * h_event, MPL_gpu_event_t * event)
+{
+    cudaError_t ret;
+    ret = cudaIpcGetEventHandle(h_event, event);
+    CUDA_ERR_CHECK(ret);
+
+  fn_exit:
+    return MPL_GPU_SUCCESS;
+  fn_fail:
+    return MPL_GPU_FAILURE;
+}
+
+int MPL_gpu_ipc_open_event_handle(MPL_gpu_event_t * event, MPL_gpu_ipc_event_handle_t * h_event)
+{
+    cudaError_t ret;
+    ret = cudaIpcOpenEventHandle(event, h_event);
+    CUDA_ERR_CHECK(ret);
+
+  fn_exit:
+    return MPL_GPU_SUCCESS;
+  fn_fail:
+    return MPL_GPU_FAILURE;
+}
+
+int MPL_gpu_event_create(MPL_gpu_event_t * event, int flag)
+{
+    cudaError_t ret;
+    ret = cudaEventCreate(event, flag);
+    CUDA_ERR_CHECK(ret);
+
+  fn_exit:
+    return MPL_GPU_SUCCESS;
+  fn_fail:
+    return MPL_GPU_FAILURE;
+}
+
+int MPL_gpu_event_sync(MPL_gpu_event_t * event)
+{
+    cudaError_t ret;
+    ret = cudaEventSynchronize(event);
+    CUDA_ERR_CHECK(ret);
+
+  fn_exit:
+    return MPL_GPU_SUCCESS;
+  fn_fail:
+    return MPL_GPU_FAILURE;
+}
+
+int MPL_gpu_event_record(MPL_gpu_event_t * event)
```

Comment:
Are you planning to add `event_query` also? The non-blocking matching part for `event_record`.

Reply:
Yes. There I am going to add it.

Reply:
@yfguo Will there be a version of `MPL_gpu_event_record` that accepts `stream` as a parameter?